fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/clientlog/KeySpaceCountTree.java

```diff
@@ -83,35 +389,181 @@ protected TupleKeyCountTree newPrefixChild(@Nonnull Object prefix) {
+                ((ResolvedRecordStoreKeyspace)indexKeyspace.getParent()).getRecordStoreKeyspace() == FDBRecordStoreKeyspace.INDEX) {
+            Index index = indexKeyspace.getIndex();
+            KeyExpression indexExpression = index.getRootExpression();
+            if (distance < indexExpression.getColumnSize()) {
+                return resolveIndexField(context, resolvedParent, object, index, indexExpression, distance);
+            }
+        }
+        return UNRESOLVED;
+    }
+
+    /**
+     * Determine whether this leaf of the {@link KeySpaceDirectory} tree is the root of a {@link com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore}.
+     * @param context an open transaction to use to read from the database
+     * @param resolvedParent the resolved parent node
+     * @param object the {@link com.apple.foundationdb.tuple.Tuple} element for this node
+     * @return {@code true} if this path stores a record store
+     */
+    protected boolean isRecordStoreLeaf(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object) {
+        return false;
+    }
+
+    /**
+     * Given a key space path for which {@link #isRecordStoreLeaf} is {@code true}, get the record store's meta-data.
+     * @param context an open transaction to use to read from the database
+     * @param resolvedParent the resolved parent node
+     * @param object the {@link com.apple.foundationdb.tuple.Tuple} element for this node
+     * @return a future that completes to the record store's meta-data or {@code null}
+     */
+    protected CompletableFuture<RecordMetaData> getRecordStoreMetaData(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object) {
+        return NO_META_DATA;
+    }
+
+    protected CompletableFuture<Resolved> resolvePrimaryKeyField(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object,
+                                                                 @Nonnull KeyExpression primaryKey, int index) {
+        List<KeyExpression> keys = primaryKey.normalizeKeyForPositions();
+        if (index < keys.size()) {
+            KeyExpression key = keys.get(index);
+            while (key instanceof NestingKeyExpression) {
+                key = ((NestingKeyExpression)key).getChild();
+            }
+            if (key instanceof FieldKeyExpression) {
+                return CompletableFuture.completedFuture(new ResolvedPrimaryKeyField(resolvedParent, ((FieldKeyExpression)key).getFieldName(), object, object));
+            }
+        }
+        return UNRESOLVED;
+    }
+
+    protected CompletableFuture<Resolved> resolveIndexField(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object,
+                                                            @Nonnull Index index, @Nonnull KeyExpression rootExpression, int fieldIndex) {
+        return UNRESOLVED;
```

Comment:
Why is this `UNRESOLVED` rather than following the same logic as in `resolvePrimaryKey`?

Reply:
Absent some new protocol in the factory / maintainer, this class would need to know for each index type how the keys it uses correspond to the root expression.

For example, a time-window leaderboard has a leaderboard index before the group starts.

Is it worth doing for, like, `VALUE`?

Reply:
I suppose there could be a method on the factory (the maintainer needs a record store and I'm not sure making one from the example path is legit) that returned a list of key expression, defaulting to normalizing the root expression. Aggregate indexes could return just the group and leaderboard could prepend a leaderboard id field.

Reply:
Yeah, I think maybe having a method on the factory is the right choice at some point. I believe there's some possibility that something like that will come out of the planner work, in order to accommodate producing some IR based on the index definition. I could also see us wanting to do `VALUE` indexes sooner rather than later, but maybe not yet.